src/Servers/Kestrel/Core/src/Internal/Certificates/CertificateConfigLoader.cs

```diff
@@ -71,7 +72,10 @@ public CertificateConfigLoader(IHostEnvironment hostEnvironment, ILogger<Kestrel
                 throw new InvalidOperationException(CoreStrings.InvalidPemKey);
             }
 
-            return (new X509Certificate2(Path.Combine(HostEnvironment.ContentRootPath, certInfo.Path!), certInfo.Password), fullChain);
+            fullChain.Import(certificatePath, certInfo.Password);
+            var certWithKey = fullChain.Single(c => c.HasPrivateKey);
```

Comment:
While I used Single() in my comment, I'm not sure if the exception text it produces is what you want for your scenario.

Previous/current/whatever behaviors:
* Empty PFX: CryptographicException
* One cert with an associated private key: Golden path
* More than one cert with an associated private key: The first cert with a private key is selected.
* No associated private keys: The first cert is selected
* It's a single X.509, not a PFX: be that cert
* It's a SignedCMS file (e.g. *.p7b):  The certificate that is associated with the first signer.

New behavior:
* Empty PFX: InvalidOperationException: Sequence contains no matching element
* One cert with an associated private key: Golden path
* More than one cert with an associated private key: InvalidOperationException: Sequence contains no matching element
* No associated private keys: InvalidOperationException: Sequence contains no matching element
* It's a single X.509, not a PFX: InvalidOperationException: Sequence contains no matching element
* It's a SignedCMS file (e.g. *.p7b): InvalidOperationException: Sequence contains no matching element

My Single() was a bit hasty, handwaving that "clearly only certs with private keys associated are relevant", but I seem to recall that SslStream has some logic where if the provided cert doesn't have a private key that it'll try looking in CU\My and/or LM\My for a matching cert to see if the key is known via X509Store.

There's also some... silliness... in private key lifetimes from a collection load vs a single cert load.  So instead of handwaving some things I'll give a bigger code splat.

I'm pretty sure this works, but I haven't run it (or, honestly, tried compiling it).  It replaces the sub-expression `new X509Certificate2(Path.Combine(HostEnvironment.ContentRootPath, certInfo.Path!), certInfo.Password)` from the "before" code.
```C#
private static X509Certificate2 LoadCertificates(
    string path,
    string password,
    X509Certificate2Collection fullChain)
{
    int count = fullChain.Count;
    X509KeyStorageFlags loadFlags = OperatingSystem.IsWindows() ?
        X509KeyStorageFlags.EphemeralKeySet : X509KeyStorageFlags.DefaultKeySet;

    fullChain.Import(path, password, loadFlags);
    bool reloadTarget = false;

    for (int i = count; i < fullChain.Count; i++)
    {
        X509Certificate2 cur = fullChain[i];

        if (cur.HasPrivateKey)
        {
            if (!reloadTarget)
            {
                // This is the target certificate, reload it with
                // non-ephemeral keys using the single-cert loader.
                reloadTarget = true;
                fullChain.RemoveAt(i);
                i--;
            }
            else
            {
                // The private key for this certificate doesn't matter,
                // replace it with a public-key-only version.
                fullChain[i] = new X509Certificate2(cur.RawDataMemory.Span);
                cur.Dispose();
            }
        }
    }

    // If we want the target reloaded, or we want the "this PFX was empty" exception
    if (reloadTarget || fullChain.Count == count)
    {
        return new X509Certificate2(path, password);
    }

    // There were no certificates with private keys, so just use the first certificate
    // and let SslStream try to find a private key later.
    X509Certificate2 target = fullChain[count];
    fullChain.RemoveAt(count);
    return target;
}
```

Reply:
So I am a little worried about this change now, it looks like we don't have any tests covering this class's functionality, we have tests that mock this config loader out, but I tried changing this code to just always return (null, null) and nothing failed...

Reply:
I don't think this is something we should try and do in rc2 given that we have very little (or no test coverage here)

Reply:
So yeah we discussed this in triage, I'll still work on this PR, but we are going to target main/NET8.0 for this work instead of trying to get it into rc2 for 7

Reply:
I'll add some tests as part of this PR so we can be confident of all the various behaviors you outlined

Reply:
I'm ok with starting in .NET 8 for sure if the change looks big. @bartonjs Maybe this is also a fancy new BCL API 😄?